Spigot-API-Patches/0171-Add-BaseComponent-array-to-a-variety-of-events.patch

```diff
@@ -0,0 +1,2360 @@
++     * Sets the kick message to display if getResult() != Result.ALLOWED
++     * <p>
++     * The specified message is NOT copied. Use {@link BaseComponentUtils#copy(BaseComponent[])} to make a copy.
++     * </p>
++     * @param components The specified message.
++     */
++    public void setKickMessageComponents(final BaseComponent[] components) {
++        this.message = null;
++        this.messageComponents = components;
++    }
++
++    /**
++     * Disallows the player from logging in, with the given reason
++     * <p>
++     * The specified message is NOT copied. Use {@link BaseComponentUtils#copy(BaseComponent[])} to make a copy.
++     * </p>
++     * @param result New result for disallowing the player
++     * @param message Kick message to display to the user
++     */
++    public void disallow(final Result result, final BaseComponent[] message) {
++        this.result = result;
++        this.setKickMessageComponents(message);
++    }
++
++    /**
++     * Returns the current kick message that will be used if getResult() != Result.ALLOWED
++     * <p>
++     * The returned value is NOT a copy. Modifying the value will result in undefined behaviour.
++     * Use {@link BaseComponentUtils#copy(BaseComponent[])} to make a copy.
++     * </p>
++     * @return The kick message to used if getResult() != Result.ALLOWED
++     */
++    public BaseComponent[] getKickMessageComponents() {
++        if (this.messageComponents == null) {
++            return this.message == null ? null : (this.messageComponents = TextComponent.fromLegacyText(this.message));
++        }
++        return this.messageComponents;
++    }
++    // Paper end
++
+     /**
+      * Gets the current result of the login, as an enum
+      *
+@@ -110,7 +157,12 @@ public class AsyncPlayerPreLoginEvent extends Event {
+      * @return Current kick message
+      */
+     public String getKickMessage() {
+-        return message;
++        // Paper start - Add BaseComponent to events
++        if (this.message == null) {
++            return this.messageComponents == null ? null : (this.message = TextComponent.toLegacyText(this.messageComponents));
++        }
++        return this.message;
++        // Paper end
+     }
+ 
+     /**
+@@ -120,6 +172,7 @@ public class AsyncPlayerPreLoginEvent extends Event {
+      */
+     public void setKickMessage(final String message) {
+         this.message = message;
++        this.messageComponents = null; // Paper - Add BaseComponent to events
+     }
+ 
+     /**
+@@ -127,7 +180,7 @@ public class AsyncPlayerPreLoginEvent extends Event {
+      */
+     public void allow() {
+         result = Result.ALLOWED;
+-        message = "";
++        this.setKickMessage(""); // Paper - Add BaseComponent to events
+     }
+ 
+     /**
+@@ -138,7 +191,7 @@ public class AsyncPlayerPreLoginEvent extends Event {
+      */
+     public void disallow(final Result result, final String message) {
+         this.result = result;
+-        thi
```

Comment:
Current BaseComponent[] broadcasting does not check permissions, while legacy text does. Figure out:

- Whether to modify the behaviour at all
- Merge broadcasting behaviour to both be permission-checking (in or out of this patch) or not permission-checking (by default anyways)
- Whether to add a permission-checking basecomponent broadcast

Also add documentation to these methods

Reply:
decided to revisit this in a later patch